### ISSUE TITLE
RFC - More tessellation control

### ISSUE THREAD

**Issue by zougloub** [openscad/openscad#4246]
`$fn`, `$fs`,  `$fa` are allowing to control the number of fragments in a shape; their goal is to control the deviation between the tessellated geometry and the ideal geometry, or to simplify rendering.

I would like to propose something which extends their functionality without changing the code too much, eg. using special values for them; it's a bit "dirty" but see, for example:

```C
if (fn == -1.0) {
  /*
    fs means max. error between polygon edge and ideal shape.
    eg. circle is tessellated using regular polygon with vertices on the circle;
    fs means max. error between polygon edge and circle;
    fa is unused.

    (partial implementation)
  */
  return int(ceil(M_PI / ( acos((r + fs) / r) )));
}
else if (0) {
  /*
   Leaving room for future $fn perversion here.
  */
}
else {
  /*
    Reuse good old way.
  */
  return Calc::get_fragments_from_r(r, fn, fs, fa);
}
```

**Comment by t-paul** [comment]
Problems with that:
* Special case for *some* places, or even more tricky to apply to other things like fonts
* Introducing magic values is usually not good
* Interferes with automatic passing along of `$fX` values if other built-ins are supposed to keep there non-magic behavior

I'm leaving a strong vote against this kind of change. There should be better ways of giving more/finer control over tesselation.

**Comment by zougloub** [comment]
I concur with your analysis @t-paul hence the RFC mention and title.

The quoted implementation of a new `get_fragments_from_r()` only handles `circle()` (and sphere/cylinder), and of course all shapes should be handled for each new convention.

I didn't formulate clearly enough that ensuing discussions would be about:
- The fact that there could be several definitions of “tessellation error”;
- Introduction of “max chord deviation considering segments on the original ideal shape” error control modality, which seems more relevant considering the goal of error control;
- Better ways than `fn/fa/fs` to control the error, vs. using magic negative values of `$fn` to encode different error modalities and up to 2 control values.

**Comment by coryrc** [comment]
Why not proposing a new special `$fe` for "means max. error between polygon edge and ideal shape".

I'm going to write this PR soon, as this is exactly what I want. I want the error to be a fair bit smaller than the line width of my printer and I don't care at all what degrees or number that is.

**Comment by coryrc** [comment]
> Why not proposing a new special $fe for "means max. error between polygon edge and ideal shape".

Heh, looking through the call stack, I see why now.

**Comment by coryrc** [comment]
I've gotten side-tracked by namespaces, but I understand the codebase far better now and will be able to execute on this.

**Comment by coryrc** [comment]
Another alternative would to allow the user to implement `get_fragments_from_r`.

The first thing I'm going to do is refactor passing a struct instead of fs, fn, fa individually to make whatever we do easier.

**Comment by coryrc** [comment]
Could this be implemented in a way which works for third-party libraries which implement non-builtins? I.e. what if someone uses polyhedron, could we intercept accesses to $fn, $fa, and $fs and provide different values? ... I don't think so.

**Comment by jordanbrown0** [comment]
> I.e. what if someone uses polyhedron, could we intercept accesses to $fn, $fa, and $fs and provide different values?

I'm not sure exactly what you're thinking of, but I have long thought that we should have a function that returns the number of segments for a circle of the specified radius/diameter, taking all of the `$f*` variables into account.

That is, `fn(r=10)` would return the number of segments that would be used for `circle(r=10)`.  What you then do with the information is up to you.

In fact, I think I implemented that recently, though I don't remember exactly why.

**Comment by jordanbrown0** [comment]
In fact, I think I implemented that recently, though I don't remember exactly why.

**Comment by coryrc** [comment]
> > I.e. what if someone uses polyhedron, could we intercept accesses to $fn, $fa, and $fs and provide different values?
> 
> I'm not sure exactly what you're thinking of, but I have long thought that we should have a function that returns the number of segments for a circle of the specified radius/diameter, taking all of the `$f*` variables into account.

Hardest part of that is picking a name.

Also, welcome advice on naming this internalcode. Today we pass `fn, fs, fa` as individual arguments and store individually all over and call `get_fragments_from_r(r, fn, this->fs, or node.fa)` in many variations, which gets replaced by a single argument/member of this class + call a function with just `r` to get the number of segments/subdivisions/fragments/what-have-you.

```
class TessellationControl
{
public:
  TessellationControl(const Parameters& parameters, const ModuleInstantiation *inst=nullptr);
  std::optional<int> circular_segments(double r) const;
  friend std::ostream& operator<<(std::ostream& stream, const TessellationControl& f);
private:
  double fn, fs, fa;
};
```

**Comment by jordanbrown0** [comment]
I would suggest `fn()`.

For the name of your class/functions... TessellationControl seems too broad; there are lots of aspects of tessellation that it doesn't control.  I would think of something that more directly brings in the words "circle", "arc", "segments", and/or "fragments".

**Comment by nophead** [comment]
It is called r2sides(r) in NopSCADlib.

**Comment by coryrc** [comment]
> I would suggest `fn()`.

Problem for me is that's also commonly a placeholder function name, would be like naming a builtin variable `i`.

> For the name of your class/functions... TessellationControl seems too broad; there are lots of aspects of tessellation that it doesn't control. I would think of something that more directly brings in the words "circle", "arc", "segments", and/or "fragments".

Hmm, yes, as I've been updating the code, the only thing $f{nas} are for what you're saying (I wasn't sure how deep it would go); it's only used in a two-dimensional context, for segments of an arc. So tessellation is indeed the wrong word.

"discretization of a circle" seems to be the more appropriate phrase? [Discretization](https://en.wikipedia.org/wiki/Discretization) at wikipedia says:

> In [applied mathematics](https://en.wikipedia.org/wiki/Applied_mathematics), discretization is the process of transferring [continuous](https://en.wikipedia.org/wiki/Continuous_function) functions, models, variables, and equations into [discrete](https://en.wiktionary.org/wiki/discrete) counterparts.

As properties of a circle, they aren't "segments" because a segment is [an area of the circle enclosed by a chord.](https://www.math-salamanders.com/parts-of-a-circle-worksheet.html). But in computer graphics it's used as shorthand for *line segments* which are what the approximation is made up of.

Using LLM eventually concludes:

```
The standard terminology is:
 Mathematical/Theoretical: Sides (or n)
 Computer Graphics/Implementation: Segments (or Line Segments)
```

(it also says 'No, the word "fragments" is not the standard, common, or authoritative term used to refer to the individual line segments that make up the polygon used to approximate a circle.')

Circles are a subset of arcs, so let's drop circle.

That brings us down to "discretization", "arc", and "segments".

```
class ArcDiscretizer {
  std::optional<int> GetSegmentCount(double r, double angle_degrees=360.0) const;
};

// Replace `double fn, fa, fs;` everywhere with:
ArcDiscretizer discretizer; // ???
```

How's this look?

**Comment by coryrc** [comment]
> Another alternative would to allow the user to implement `get_fragments_from_r`.

Only problem I've hit with this idea is the CSG intermediate format. It would be a big project to allow this.

**Comment by coryrc** [comment]
Problem with refactoring so most of the code doesn't need to know what parameters control discretization:

Python functions like `python_circle` are doing:

```
  char *kwlist[] = {"r", "d", "fn", "fa", "fs", NULL};

  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|ddddd", kwlist, &r, &d, &fn, &fa, &fs)) {
```

to get out fn, fa, and fs as args, so every function has to know about every possible option because the above is written to allow in Python: `circle(1.2, 2.4, 1, 2, 3)` to set all of r, d, fn, fa, and fs. Which, personally, seems bad: specifying both r and d is not good, but, relevantly, fn/fa/fs should have been keyword-only. I can make that latter change to make them keyword-only and then only `ArcDiscretizer` needs to know about the params. How does this sound?

 [Docs for PyArg_ParseTupleAndKeywords](https://docs.python.org/3/c-api/arg.html#c.PyArg_ParseTupleAndKeywords),

**Comment by coryrc** [comment]
@gusohler do you have an opinion on making fn/fa/fs (and possible future discretization controls) keyword-only args?

(details in previous post)

**Comment by jordanbrown0** [comment]
Yes, I agree that `fn()` is awfully general.  But it's a **really** good match for the existing practice.

For general use, I'd say that "discretization" is too non-common-English.  But for internal use, shrug.

$fn/$fs/$fa **are** used to control spheres, so they are at least somewhat used in 3D contexts.  And I expect that if we ever did Bézier patches those variables would have a role, though it wouldn't be as easily defined as the role for circular stuff.

For CSG output, there needs to be an interface back out that turns your object back into a list of input variables - perhaps as a string holding a comma-separated list, or perhaps even via an `operator<<` to a stream.

To be clear: I have no opinion on how this fits into the Python API; I haven't looked at it.

**Comment by jordanbrown0** [comment]
To be clear:  I have no opinion on how this fits into the Python API; I haven't looked at it.

**Comment by coryrc** [comment]
> $fn/$fs/$fa **are** used to control spheres, so they are at least somewhat used in 3D contexts.

I meant that `get_fragments_from_r` is only used for calculating segments of 2D parts; in the sphere, its result is used only for the polygon of the sphere intersecting the Z plane. I would expect a "3D" version of this function to instead return the number of desired facets instead, but instead we get these weird jumps:

| $fn | faces | triangles |
|-----|----| ------|
| 3 | 5 | 8 |
| 4 | 6 | 12 |
| 5 | 12 | 26 |
| 6 | 14 | 32 |
| 7 | 23 | 50 |

However, there are some other "truly 3D" segmentation calculations in calc.cc like `get_diagonal_slices` and those could be considered tessellation? but they still return a one-dimensional answer so they also could be considered just segmentation.

> For CSG output, there needs to be an interface back out that turns your object back into a list of input variables - perhaps as a string holding a comma-separated list, or perhaps even via an `operator<<` to a stream.

It's already the latter, which writes out `$fn=.., $fa=..., $fs=...` every time, so I was thinking "oh no, but then everything will need to know how to process them" then... head-smack: we only need to and should write `$fn`! So a user function is fine.

But I also realized that segment count doesn't recalculate if you scale a shape. Not that if you'd asked me before, I would answer wrong, but it never occurred to think about before.

**Comment by jordanbrown0** [comment]
You can't *quite* only write out $fn, because it looks like there are a couple of cases (notably in extrusion) where it does some calculation based on the size of the child geometry, which isn't knowable at CSG time.  The `node.fa` reference there was a give-away.

But CSG is just a subset of OpenSCAD-language, so the fact that "everything" needs to know how to process them doesn't really mean much - they already need to know.  Theoretically we might reduce to $fn for the cases where we can, and then a CSG consumer might not need to know as much as a full OpenSCAD-language processor does, but it looks like there are still a few cases where it does.

**Comment by jordanbrown0** [comment]
I was curious about the number of faces on the spheres.  What's directly calculated is the number of vertices.  It's easiest for me to think of it as lines of latitude and longitude.  There are $fn lines of longitude, so the equator has the same number of segments as a $fn circle.  There are $fn/2, rounded up, lines of latitude, because you only walk across half of the circle.  The number of vertices is the product of those two, so (for instance) $fn=7 yields 7 lines of longitude times 4 lines of latitude is 28 vertices.

But there are only faces **between** the lines of latitude, so that's one fewer ring, so it's 7 times 3 is 21, plus the two polar faces is 23.

**Comment by coryrc** [comment]
> You can't _quite_ only write out $fn, because it looks like there are a couple of cases (notably in extrusion) where it does some calculation based on the size of the child geometry, which isn't knowable at CSG time.

Ah, darnit, I hadn't gotten that far yet, am still at the trivial cases. You're talking about `extrudePolygon()` in `linear_extrude.cc`.

> But CSG is just a subset of OpenSCAD-language, so the fact that "everything" needs to know how to process them doesn't really mean much - they already need to know. Theoretically we might reduce to $fn for the cases where we can, and then a CSG consumer might not need to know as much as a full OpenSCAD-language processor does, but it looks like there are still a few cases where it does.

It's my understanding it's used as the "lossless" export to FreeCAD and maybe others in the future, so if I add the original request of this Issue -- i.e. set segment count so error is just below some target -- they may need to be updated.

Except... dur, the point is FreeCAD *doesn't* need to care about these parameters because it can represent the true shape? (except when circle($fn=) is used for polygons?) So it should be fine for it to ignore any new segmentation specials?

So for the purposes of this Issue I think it's as simple as writing out `$fe` everywhere `$fs` is written for CSG?

**Comment by MichaelAtOz** [comment]
The [wiki section ](https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/Other_Language_Features#Circle_resolution:_$fa,_$fs,_and_$fn)is peppered with the word 'fragments', so regardless of LLM OpenSCAD speak should perhaps be **fragments(r)** (?) But...
The wiki also has:
 "Or you can embed this OpenSCAD version in your code to work out what's going on, you need to set r= to your size
   `echo(n=($fn>0?($fn>=3?$fn:3):ceil(max(min(360/$fa,r*2*PI/$fs),5))),a_based=360/$fa,s_based=r*2*PI/$fs);`"

Which IIRC I added, does it really need a built-in?

**Comment by coryrc** [comment]
> does it really need a built-in?

I proposing to change it, so everybody with a copy will need to update, so why not have it be available as a builtin? They're trivially shadowed, so it's not like it'll cause problems.

**Comment by jordanbrown0** [comment]
CSG rules are murky.  The only thing that's truly promised is that a current OpenSCAD can read its own CSG output, and probably that it can read all previous OpenSCADs' output.  Emitting $fe in place of previous values - if you can even figure out what the equivalent is - would definitely break reading the CSG file on all existing consumers, or at least all existing ones that do circles right.

The rationale for adding a built-in is so that we can tweak the algorithm and have programs able to cope without having to be upgraded.  (Past needing to use the new function, of course.)  Why would we want to tweak the algorithm?  To add new ways to control the accuracy, as discussed here.  To improve it, e.g. to force $fa/$fs to generate multiples of four, so that the circle always touches all four sides of the bounding box.

**Comment by MichaelAtOz** [comment]
Hmmm, I think I need to add int ie floor()...done wiki updated.

**Comment by MichaelAtOz** [comment]
Damn markup that is r * 2 * PI

`echo(n=($fn>0?($fn>=3?$fn:3):ceil(max(min(360/$fa,r*2*PI/$fs),5))),a_based=floor(360/$fa),s_based=floor(r*2*PI/$fs));`

**Comment by coryrc** [comment]
> CSG rules are murky. The only thing that's truly promised is that a current OpenSCAD can read its own CSG output, and probably that it can read all previous OpenSCADs' output. Emitting $fe in place of previous values

Not "in place" -- it's an additional term used for calculating the number of segments. If you read a new file with an old OpenSCAD, you could get a different number of segments, but a new OpenSCAD would read old and new files correctly. It *could* be done by adding a line to `get_fragments_from_r`:

```
  if (r < GRID_FINE || std::isinf(fn) || std::isnan(fn)) return 3;
  if (fn > 0.0) return static_cast<int>(fn >= 3 ? fn : 3);
  if (fe > 0.0) return static_case<int>(ceil(M_PI / ( acos((r + fs) / r) ))); // <== NEW LINE
  return static_cast<int>(ceil(fmax(fmin(360.0 / fa, r * 2 * M_PI / fs), 5)));
```

But then there's so many callers to update, so I think it makes sense to refactor so the callers don't need to know about every control variable. (also haven't looked whether that formula is correct, just copy-pasted from the first post)

**Comment by jordanbrown0** [comment]
Sorry, I read a previous comment as writing $fe instead of $fs, when you meant that anywhere we write $fs we should **also** write $fe.

I also realized that we **must not** simply calculate and emit $fn, as appealing as that might be.  If you do that, the consumer can't tell whether or not the original had $fn set.  A reasonable interpretation would be that an explicit $fn requests a regular polygon while no explicit $fn requests a circle, and always emitting $fn would break that interpretation.

**Comment by coryrc** [comment]
> I also realized that we **must not** simply calculate and emit $fn, as appealing as that might be. If you do that, the consumer can't tell whether or not the original had $fn set. A reasonable interpretation would be that an explicit $fn requests a regular polygon while no explicit $fn requests a circle, and always emitting $fn would break that interpretation.

Good point, the same effect could be produced by outputting values for $fa and $fs which would have the similar effect in older versions of Openscad as the proposed `$fe` would in newer versions.

**Comment by coryrc** [comment]
> The [wiki section ](https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/Other_Language_Features#Circle_resolution:_$fa,_$fs,_and_$fn)is peppered with the word 'fragments',

I've proposed a change to the page to fix that.

**Comment by coryrc** [comment]
As stated before, I want to be able to say "make sure the tessellated shape is no further from the platonic ideal by X mm". I plan to implement it first for just the things which were using `get_fragments_from_r`: `circle`, `cylinder`, `sphere`, `offset`, `rotate_extrude`, imports of SVG, and `text`. (Theoretically we could make it affect DXF data import, but currently that's fixed at 36 segments per circle).

Here are some implementation details to decide on:

1. The name `$fe`
2. What exactly `$fe` is measuring
3. How it interacts with `$fs` and `$fa`
4. Whether it should interact with `$fn`


Regarding `#`2, a discretized circle currently inscribes the platonic circle. I think if `$fe` is the controlling factor, then the distance from the midpoint of a polygon segment to the circumference of the platonic circle should the target (i.e. \#segments is the lowest possible while keeping that measurement less than or equal to `$fe`), except the minimum number of segments is 5 for a whole circle.

Regarding `#`3, I think it should override `$fs` and `$fa`, except for honoring the builtin minimums for those values. So segments will never be smaller than 0.01 and no more than 36000 segments per circle. When `$fe` is zero, it has no effect.

Regarding `#`4, I think `$fn` should override `$fe` just like it overrides the others.

**Comment by coryrc** [comment]
I have a number of AIs in [#6251](https://github.com/openscad/openscad/pull/6251) to work on, have been working on color instead. Still will be working on this.

**Comment by coryrc** [comment]
Color spec is done, coming back to this soon.

**Comment by coryrc** [comment]
After the very big refactor, adding `$fe` was really easy: See #6361. (EDIT: well, at least the part I did so far...)

**Comment by coryrc** [comment]
Regarding the behaviour of having `$fe`, `$fa`, and `$fs` set:

Before this change, we pick `min(fa_segments,fs_segments)`. Currently I have `fe ? fe_segments : min(fa_segments,fs_segments)`, because if you've set a tolerance, I don't think there's any reason to be more precise. However, I could see someone disagreeing, and then we could instead choose `max(fe_segments, min(fa_segments,fs_segments))`.

Does anyone think $fe's calculated segments should instead be a minimum number of segments in conjunction with $fs/$fa calculations?

**Comment by jordanbrown0** [comment]
No strong feeling.

Two things to think about:
* $fa and $fs have default values that yield relatively low numbers of sides.  That means that if the rule is min(fe_segs, fa_segs, fs_segs) and you want to control it with $fe, you *must* set very small values for $fa and $fs.
* Explainability is critical.  People already have trouble understanding how $fa and $fs interact.

**Comment by coryrc** [comment]
In #6366 I'm trying to document why and how much untwisted linear_extrude needs discretization. There's pictures and code there if you want the long version, but I think the error can be measured by this algorithm:

```
for ( a given slice count ) {
    for ( every vertex? just the furthest one? any!? ) {
      Pick the center slice or adjacent one
      Get the formula for the line from this vertex on low end of the slice to adjacent vertex on the top end of the slice.
      Invent a new vertex half-way between the two vertices on the original polygon.
      Get the formula for that new vertex projection
      Calculate the xy-difference between the two formulas when z is at the center of the chosen slice.
      error=max(error,calculation)
    }
}
```

Hopefully that can be simplified to not have any loops when starting with $fe.

**Comment by coryrc** [comment]
In scaled untwisted linear extrude, you have the scenario where the furthest vertex has close neighboring vertices, but a closer vertex has neighbors far away, and so the error would be higher in the latter case. So if one vertex is even possible to know, it's not necessarily the furthest vertex.

```
module sh() polygon([[10,0],[8,-2],[-2,0],[8,2]]);
//module sh() scale([1,0.1]) rotate(45) translate([6,6,0]) square(10, center=true);

module s(s) linear_extrude(h=30, scale=[1/20,20], slices=s) sh();

steps=10;

for(i=[0:9]) {
  $t=i/10;
  
  translate([0,0,i]) linear_extrude(h=1) projection(cut=true)
  {
    low_slices=($t*steps)+1;
    translate([0,0,low_slices%2==1? -15 : -15+30/low_slices/2])
      difference() { s(40); s(low_slices);}
  }
}
```

<img width="904" height="549" alt="Image" src="https://github.com/user-attachments/assets/130914c8-cf49-4a1e-a518-1af6f35d2c75" />

**Comment by coryrc** [comment]
(BTW, if anybody already knows wtf this is and if it already has a clever solution, please LMK)

Here's an interesting error measurement:

```
module sh() polygon([[10,0],[8,-2],[-2,0],[8,2]]);

module s(s) linear_extrude(h=30, scale=[1/20,20], slices=s) sh();

slices=5;

for(i=[1:slices]) {  
  translate([0,0,i]) linear_extrude(h=1) projection(cut=true)
  {
    height=30/slices;
    translate([0,0,-((1/2)*height+(i-1)*height)])
      difference() { s(40); s(slices);}
  }
}
```

<img width="904" height="549" alt="Image" src="https://github.com/user-attachments/assets/2a44068a-475f-4b5c-9cb5-004bdfcc7f2d" />

What we're looking at is the error for each of 5 slices at half the slice height (there's a glitch on the top one; it's not actually that wide). The sides adjacent to the vertex closest to the origin have the largest error on the bottom slice, approximately twice the error of the center slice. I'm not certain it's possible to measure error without an iterative solution and scanning every vertex.

**Comment by coryrc** [comment]
Okay, updating linear_extrude to use $fe is not going to be simple. In short, linear_extrude modifies two variables: the number of horizontal slices and the number of vertices in the input polygon. Currently it calculates horizontal slices first and then calculates how many vertices (if any) to add. The error term decreases when incrementing either value, but the marginal decrease falls as value increases. Say you have two linear_extrude with the same parameters but different children:

*  A polygon with 100 evenly-spaced vertices (f.ex. a regular 100-gon)
* A polygon with 100 vertices, but mostly concentrated in groups (f.ex. a square with rounded corners)

The ROI on adding vertices is much higher for the second polygon, because you're splitting very large sides; while the former requires 100 new vertices before any reduction in maximum error. So the first polygon will optimally have more slices and few/none added vertices. The second polygon would require many, many more slices without adding vertices. But this requires completely restructuring how linear_extrude calculates the two values.

Hmm, actually this is starting to sound like a leetcode hard.

When $fe is set, ignore the existing logic. First check for any explicit settings (slices=, segments=); if both, you're done; if one is set, then you can search for the second. But the most fun one is if neither is fixed. The cost function could be number of vertices. Then you're searching for the pair of values with the lowest cost where the max error <= $fe.

I'll make it even more fun: variable-height slices. As shown in my previous examples, some slices can have much more error than others; maximum error can be reduced without additional vertices nor slices by redistributing the share of height across the slices.

**Comment by coryrc** [comment]
Oh shit, it's worse: NP-hard. And my research is coming up with terms I've seen floating around but was ignorant of.

* Hausdorff Distance
* Mesh Decimation
* Quadric Error Metrics (QEM)
* Edge Collapse

More research to come!

**Comment by coryrc** [comment]
> I'll make it even more fun: variable-height slices. As shown in my previous examples, some slices can have much more error than others; maximum error can be reduced without additional vertices nor slices by redistributing the share of height across the slices.

I could use the same heuristic here as is used for adding vertices: subdivide the slice with the most error next instead of evenly cutting slices initially. You'd still really want to be able to switch between adding slices or adding vertices and not have fixed slices up-front.

**Comment by coryrc** [comment]
More terms:

* Ruled Surface Approximation
* The curved faces created by the twist are Hyperbolic Paraboloids.
* Anisotropic Meshing
* Recursive Bisection / Adaptive Subdivision

Now to look at Manifold to see what operations it provides.

**Comment by coryrc** [comment]
> Now to look at Manifold to see what operations it provides.

This doesn't seem to be the kind of thing it does, though there are many terms I am ignorant of.

**Comment by coryrc** [comment]
Given the other stuff could take a while, I have #6361 working which adds $fe for circles, spheres, and cylinders. It is ready for review.

**Comment by jordanbrown0** [comment]
I would be a little concerned that somebody might expect to set `$fe` to control curve resolution, and then be surprised when it *doesn't* control some of the cases.  I don't think it needs to do exactly the right thing; I think it just needs to do something that's sort of reasonable.

Ref #6059, my opinion (which is not universally shared) is that the intent of $fa/$fs/$fe is to produce an approximation of a circle, and the details of that approximation are not all that important.  ($fn is a different animal; at least for the circular stuff it needs to produce exactly what's specified.)

**Comment by coryrc** [comment]
> I would be a little concerned that somebody might expect to set `$fe` to control curve resolution, and then be surprised when it _doesn't_ control some of the cases.

I didn't say I was done! I'm just not putting it in the first PR.

Worse-case it seems I can do it non-optimally in O(n) time where n is the final number of slices by greedily setting a slice height matching the desired error and not adding any vertices. But "Recursive Bisection" looks like it should be O(v ln v) (where v is the final number of vertices), optimal, and not hard to implement. Only the generic "tessellate any surface" problem is NP-hard; this appears to have a polynomial solution.

**Comment by coryrc** [comment]
@kintel do you think we can add the proposed $fe? Do you want to hold out for a complete implementation or are you okay gradually committing it? (Right now it works for arcs; linear and rotate extrude are in development)

**Comment by coryrc** [comment]
Still waiting on review.

**Comment by coryrc** [comment]
The assignment bot is kinda broken, but I am still working on this. My PR was accepted after I updated it to put $fe behind a feature.

As I said before, *efficiently* modeling a twisted and scaled linear extrusion is quite difficult... I'm currently working on this in Manifold directly ([discussion](https://github.com/elalish/manifold/discussions/1444#discussioncomment-15725902), no code shared yet). I currently have it incompletely working with twist but no scaling:

<img width="1000" height="600" alt="image" src="https://github.com/user-attachments/assets/379ca547-8acc-4888-a705-494e3e3ec324" />

Where the vertices are near the axis, it needs fewer to approximate the curve. So instead of every "spine" needing 8 vertices and taking 32 vertices, I got it down to 27 with a simplistic algorithm. Sadly also a wrong one, as the diagonals aren't being checked for error. Hoo boy every time I think I can progress I realize there's something I missed...

**Comment by coryrc** [comment]
My algorithm is missing some planned optimizations, but I currently have:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/0a9da497-7687-4c8a-a120-6788ceb24dc3" />

That's 10 units tall, offset square twisted 60 degrees and almost every single surface you can see is confirmed to be within `0.1`. The rightmost edge is closer to the origin, so it only needs to be 3 steps tall, while the other two need 5 steps. The row grows in height if it's more efficient to be taller, and vice versa. We start at 5 rows tall and end with 3 rows on the right half, so we have fewer vertices and faces to represent the same accuracy.

**Comment by jordanbrown0** [comment]
My take is that this is mostly done (in PR #6361), and that what's missing is fine tuning some linear extrusion cases.  @coryrc , perhaps you could close this issue out as fixed and create a new one?

**Comment by coryrc** [comment]
> My take is that this is mostly done (in PR #6361), and that what's missing is fine tuning some linear extrusion cases.  @coryrc , perhaps you could close this issue out as fixed and create a new one?

It isn't and it's proving pretty tricky to do a good job with the extrusions. Currently they have no effect from $fe.

In thinking I need to get it doing a bad job with $fe first and it can be improved over time.

**Comment by coryrc** [comment]
I got something, well, many somethings that work well with no scaling or small scaling, but with extreme scaling ratio my heuristics all fail.

I'm going to try something dumb next, will result in far more faces than optimal, but at least should be acceptably cheap to calculate and implement.

**Comment by jordanbrown0** [comment]
Maybe "fine tuning" was a poor choice of words. The current status is that $fe works for everything but linear extrusion, right? Could we close this one and open one that's specific to linear extrusion?

**Comment by coryrc** [comment]
> The current status is that $fe works for everything but linear extrusion, right?

I don't understand the desire to close this issue without finishing it. It'll still just be an open issue. It doesn't work with any extrusion nor roof.

However, now that I finally understand the problem space just enough, I knew what to search for and found and implemented Rivara’s Longest-Edge Bisection and it looks good:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/62b92daa-4538-4009-b83f-26fa7542f10b" />

So I think I can pull it together soon. Also there might be other implementations already suitable for use, i.e. https://github.com/svenpilz/openmesh-local-refinement ?

**Comment by coryrc** [comment]
In the non-scaled case, it yields more faces than my hack job, but it works for all cases, and improvements could always be made later.

**Comment by jordanbrown0** [comment]
I'd like to replace the very general topic "more tesselation control" with a more specific one, so that it's easier to understand the list of open issues.

**Comment by coryrc** [comment]
Yeah, that's fine.

I also have realized a mistake I made: once scaling is involved, it's no longer a ruled surface. Which turns measuring the greatest error into a search.

But it does get close, so right now I've got:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/9b92542c-117e-4fbe-8562-6bdfb2beb8ca" />

**Comment by coryrc** [comment]
No, I was incorrect in the last comment. It's still a ruled surface because the scaling and twist is fixed per z value. Lined tracing the surface which aren't horizontal can be curved, but it's still a ruled surface. If they weren't fixed for a given z, then there could be a problem.

**Comment by coryrc** [comment]
Either way, I can write a few tests using a search and just midpoints and check for deviance. Or maybe a mathematician will come across this and take pity on me!

**Comment by coryrc** [comment]
I don't know where I got the idea you could just check midpoints of edges; maybe with unity scaling? Anyway, here's visual proof you can't:

<img width="652" height="371" alt="Image" src="https://github.com/user-attachments/assets/c26456b1-24e3-4643-9e28-41df31b80042" />

I picked two points on the same surface, made a cylinder between them, and gradually increased the diameter of the cylinder. If the midpoint was the highest error, the yellow at the midpoint would be the last thing obscuring the cylinder. It's obviously not at the midpoint.

[ManifoldCAD.org proof](https://manifoldcad.org/#Proof%20of%20non-midpoint%20max%20error%3Ccode%3Eimport%20%7B%20Mesh%2C%20Manifold%2C%20GLTFNode%2C%20getGLTFNodes%2C%20setMaterial%20%7D%20from%20'manifold-3d%2FmanifoldCAD'%3B%0A%0A%2F%2F%201.%20Existing%20extruded%20shape%0Aconst%20s4%20%3D%20%5B%5B-1%2C%20-0.5%5D%2C%20%5B2%2C%20-0.5%5D%2C%20%5B2%2C%205%5D%2C%20%5B-1%2C%205%5D%5D%3B%0Aconst%20ex%20%3D%20Manifold.extrude(s4%2C%2010%2C%20300%2C%20180%2C%20%5B3%2C%200.5%5D)%3B%0A%0A%2F%2F%202.%20Define%20the%20two%20endpoints%20for%20the%20cylinder%0Aconst%20pA%20%3D%20%5B1%2C%200%2C%205%5D%3B%0Aconst%20pB%20%3D%20%5B-4.5%2C%200.25%2C%2010%5D%3B%0A%0A%2F%2F%20Calculate%20direction%20vector%20and%20distance%20(height)%0Aconst%20dir%20%3D%20%5BpB%5B0%5D%20-%20pA%5B0%5D%2C%20pB%5B1%5D%20-%20pA%5B1%5D%2C%20pB%5B2%5D%20-%20pA%5B2%5D%5D%3B%0Aconst%20height%20%3D%20Math.sqrt(dir%5B0%5D*dir%5B0%5D%20%2B%20dir%5B1%5D*dir%5B1%5D%20%2B%20dir%5B2%5D*dir%5B2%5D)%3B%0A%0A%2F%2F%20Normalize%20the%20direction%20vector%20to%20become%20our%20Z-axis%20(length%20%3D%201)%0Aconst%20zAxis%20%3D%20%5Bdir%5B0%5D%20%2F%20height%2C%20dir%5B1%5D%20%2F%20height%2C%20dir%5B2%5D%20%2F%20height%5D%3B%0A%0A%2F%2F%20Choose%20a%20temporary%20vector%20that%20isn't%20parallel%20to%20zAxis%20to%20calculate%20the%20X-axis%0Aconst%20temp%20%3D%20Math.abs(zAxis%5B2%5D)%20%3C%200.9%20%3F%20%5B0%2C%200%2C%201%5D%20%3A%20%5B1%2C%200%2C%200%5D%3B%0A%0A%2F%2F%20X-axis%20%3D%20temp%20x%20zAxis%0Aconst%20xAxisRaw%20%3D%20%5B%0A%20%20%20%20temp%5B1%5D%20*%20zAxis%5B2%5D%20-%20temp%5B2%5D%20*%20zAxis%5B1%5D%2C%0A%20%20%20%20temp%5B2%5D%20*%20zAxis%5B0%5D%20-%20temp%5B0%5D%20*%20zAxis%5B2%5D%2C%0A%20%20%20%20temp%5B0%5D%20*%20zAxis%5B1%5D%20-%20temp%5B1%5D%20*%20zAxis%5B0%5D%0A%5D%3B%0Aconst%20xLen%20%3D%20Math.sqrt(xAxisRaw%5B0%5D*xAxisRaw%5B0%5D%20%2B%20xAxisRaw%5B1%5D*xAxisRaw%5B1%5D%20%2B%20xAxisRaw%5B2%5D*xAxisRaw%5B2%5D)%3B%0Aconst%20xAxis%20%3D%20%5BxAxisRaw%5B0%5D%20%2F%20xLen%2C%20xAxisRaw%5B1%5D%20%2F%20xLen%2C%20xAxisRaw%5B2%5D%20%2F%20xLen%5D%3B%0A%0A%2F%2F%20Y-axis%20%3D%20zAxis%20x%20xAxis%0Aconst%20yAxis%20%3D%20%5B%0A%20%20%20%20zAxis%5B1%5D%20*%20xAxis%5B2%5D%20-%20zAxis%5B2%5D%20*%20xAxis%5B1%5D%2C%0A%20%20%20%20zAxis%5B2%5D%20*%20xAxis%5B0%5D%20-%20zAxis%5B0%5D%20*%20xAxis%5B2%5D%2C%0A%20%20%20%20zAxis%5B0%5D%20*%20xAxis%5B1%5D%20-%20zAxis%5B1%5D%20*%20xAxis%5B0%5D%0A%5D%3B%0A%0A%2F%2F%203.%20Native%20Manifold%20Mat4%20Array%20(16-element%20flat%20array%2C%20Column-Major%20Layout)%0A%2F%2F%20Column%201%3A%20xAxis%20%7C%20Column%202%3A%20yAxis%20%7C%20Column%203%3A%20zAxis%20%7C%20Column%204%3A%20Translation%20(pA)%0Aconst%20transformMatrix%20%3D%20%5B%0A%20%20%20%20xAxis%5B0%5D%2C%20xAxis%5B1%5D%2C%20xAxis%5B2%5D%2C%200%2C%20%2F%2F%20Column%201%20(X%20basis%20direction)%0A%20%20%20%20yAxis%5B0%5D%2C%20yAxis%5B1%5D%2C%20yAxis%5B2%5D%2C%200%2C%20%2F%2F%20Column%202%20(Y%20basis%20direction)%0A%20%20%20%20zAxis%5B0%5D%2C%20zAxis%5B1%5D%2C%20zAxis%5B2%5D%2C%200%2C%20%2F%2F%20Column%203%20(Z%20basis%20direction)%0A%20%20%20%20pA%5B0%5D%2C%20%20%20%20pA%5B1%5D%2C%20%20%20%20pA%5B2%5D%2C%20%20%20%201%20%20%2F%2F%20Column%204%20(Translation%20vector%20%26%20W%20component)%0A%5D%3B%0A%0A%2F%2F%20Generate%20the%20base%20cylinder%20pointing%20along%20the%20Z-axis%20(length%20matches%20distance%20exactly)%0Aconst%20h%20%3D%200.8%3B%0Aconst%20baseCylinder%20%3D%20Manifold.cylinder(height%2C%20h%2C%20h%2C%2032%2C%20false)%3B%0A%0A%2F%2F%20Orient%20and%20place%20the%20cylinder%20perfectly%20between%20pA%20and%20pB%0Aconst%20orientedCylinder%20%3D%20baseCylinder.transform(transformMatrix)%3B%0A%0A%2F%2F%204.%20Color%20the%20cylinder%20red%0Aconst%20redCylinder%20%3D%20setMaterial(orientedCylinder%2C%20%7B%0A%20%20%20%20baseColorFactor%3A%20%5B1.0%2C%200.0%2C%200.0%2C%201.0%5D%20%0A%7D)%3B%0A%0Aconst%20blueSphere%20%3D%20setMaterial(Manifold.sphere(.3%2C%2020)%2C%20%7B%0A%20%20%20%20baseColorFactor%3A%20%5B0.0%2C%200.0%2C%201.0%2C%201.0%5D%20%0A%7D)%3B%0A%0A%2F%2F%205.%20Union%20the%20shapes%0Aconst%20combinedResult%20%3D%20ex.add(redCylinder).add(blueSphere.translate(pA)).add(blueSphere.translate(pB))%3B%0A%0Aexport%20default%20combinedResult%3B)

**Comment by coryrc** [comment]
Okay, so Emmett (of Manifold fame) [pointed out to me: `There's also things like normals variation and triangle quality - think Delaunay condition`](https://github.com/elalish/manifold/discussions/1444#discussioncomment-17207050), so even if it was feasible to come up with the optimal solution, it wouldn't be what we actually want.

Anyway, I'm thinking I'm going to make a library with two APIs and at least an implementation of both:

* linear_extrude, take in polygons and $fe and break it into individual outlines of ruled surfaces from each edge of the polygon, concurrently calling the next function on each
* take a ruled surface outline and triangulate it, given error tolerance

I'm going to do something ... *mostly* correct, by sampling the error of each triangle at a few points and plugging that into my existing, working algorithm. Then I and/or others can improve or add new algorithms ([i.e. there appear to be algorithms designed to help](https://www.semanticscholar.org/paper/Simplification-of-surface-mesh-using-Hausdorff-Borouchaki-Frey/5d1c4af3017e955543e90fc6eea2403feaa7ab4c)).

**Comment by jordanbrown0** [comment]
Documentation should probably explicitly say that the exact results are subject to change without notice.

**Comment by coryrc** [comment]
For sure, but I think they'll want it fixed to come out of experimental. 

For the future, I think the only way to choose alternative algorithms would be another $special, if it was desired. Well, if I wasn't sure about which to use, I suppose I could do it now for people to compare alternatives.

One reason we might want to keep alternatives is the time cost versus accuracy trade-off, but also I'm planning an algorithm to push vertices off the surface, but if somebody tries to extrude very thin sections (thinner than 2*error) or acute angles, it could be made non-manifold; a similar problem already exists though (high twist) so it's not a showstopper without workaround.

I've found some papers which describe mesh algorithms for general ruled surfaces with tolerance. I'll implement at least one.

I think every surface can be split into surfaces which are either entirely concave or convex. I'm not yet sure if this could require more than two. I think the inflection curve is the closest point on each rule closest to the origin. Without scaling this is the same point along the segment being projected, but with scaling it moves; if it can move off the segment and then back on, this would make multiple surfaces per segment, which would be annoying. Anyway, if I could make a surface be entirely convex or concave, it'd be cheaper to calculate where to put your vertex off the surface (it would always be on S(u,v)+epsilon).

**Comment by jordanbrown0** [comment]
> I think the only way to choose alternative algorithms would be another $special

I would draw a distinction between an alternative and fine-tuning the behavior.

An alternative, that accepted materially different inputs or produced materially different results, where there's a good reason to retain both behaviors, yes, it'd need independent control. (Though this picture is already complex enough; there better be a really good reason for adding another scheme.)

Fine-tuning, small changes that would accept basically the same inputs and return basically the same results, only in some sense better, I'd just replace the old with the new.

(As an example... were it not for the fact that we've published the algorithm used for $fa and $fs, I wouldn't hesitate to change that algorithm so that it always produced a multiple of four sides so that the vertexes are axis-aligned.  But we've published the algorithm and people have duplicated it, and backing away from that commitment is difficult.)

**Comment by coryrc** [comment]
So the reason I'm thinking of future alternative options is when you may want to guarantee a dimension, something trivial with a cylinder but not with this! The normal algorithm may not meet it perfectly, but you may not want to run the most precise option because it takes 5 minutes to run instead of 3 seconds (just making up numbers here, but remember, the optimal solution is NP, so this is not crazy). And that 5 minutes could be because of this algorithm or because so many extra vertices and faces are used, not sure. But this is for edge cases... I'm thinking more of complicated polygons with weird scaling; anything with scaling less than order-of-magnitude difference probably will be fine, and I think twist mostly doesn't matter because once you reach very high twist, the twist rate limit pops in to limit any given triangle's reach.

So I'm just thinking of having an escape hatch for people that get to "hey I modeled this complicated thing and it's almost perfect but it still violates the constraint I set in a few places! Is there anything I can do besides redo all my work again or set $fn to a billion and wait a week for it to render but can't be machined/3d-printed because it's too complex?" Because if the error peaks are between where I plan to sample *and* really spiky, 99.9% of the surface can be within the error tolerance but tiny pieces can stick out with no other way to remove, which would be really annoying. But we're talking something >99% of users would never touch, I'm not going to implement it as a final feature -- it would be future work.

**Comment by coryrc** [comment]
Oh no, apparently building a mesh is such a big problem in CFD that mesh approximation of an arbitrary surface can be 60% of the computational cost of doing CFD... and I remember CFD being a user of Beowulf clusters back in the day because it was so demanding...

I don't think anyone involved in this thread realized we're asking for something on the cutting edge of research and with a massive computational cost.

OTOH it's not an arbitrary surface, but I doubt anyone has already researched our exact problem, and I've noticed a few ways to make things faster, but I don't feel so bad about my naive attempts anymore!

One of the reasons is just measuring error for a single point is expensive, and, well, you have to do that a lot because the differential equations can't be solved algebraically. Rather than search the triangle for the point furthest from the surface (each one requiring you search the surface for the point closest to it), I'm going to pick a handful of points most likely to be the furthest and only use those (and I'll discount the error knowing this), but overall it's likely small portions of the mesh will violate constraints. Additionally in most cases the surface will be known to be concave or convex, and I think I can use this to converge double logarithmically.

So for the first implementation, the promise will definitely need to be scaled back to "the maximum error will be about $fe and the exact triangulation will not be stable across versions". (The latter part I'm thinking about the listserv discussion about sphere algorithms and cylinder unions)

**Comment by coryrc** [comment]
<img width="681" height="628" alt="Image" src="https://github.com/user-attachments/assets/deb9625e-4a3d-403b-a096-0465c09fdd0f" />

Progress with my greatly-cleaned-up library. Visible at https://github.com/coryrc/SurfaceMeshing if you wish to follow along.

**Comment by coryrc** [comment]
Every triangle on that surface was generated automatically. The midpoint of every edge and 16 sample points fairly-evenly spread across the triangle's interior have their Hausdorff distance measured with Newton-Raphsom to find the closest point on the true surface. Every triangle passes the test.

**Comment by coryrc** [comment]
Same deal, 360 degree rotated extrusion:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/a5739b05-0fd0-4218-8148-d20a014ff3fa" />

However, I'm still getting failure to converge when cranking the scale differential way up. For example 360 degrees, {20, 0.6} scale, I see the triangle error seems to be rising over time... I haven't put in an explicit twist limit yet, so that might be a cause.

(this is all with $fe=0.1)

**Comment by coryrc** [comment]
Actually... I'm going to try to get this done with the algorithm as it is. Looking at the output, it does not seem to be close to the [Delaunay condition](https://en.wikipedia.org/wiki/Delaunay_triangulation), there are several slivers to be found. That's not mandatory, but it is a sign I should probably use better algorithms rather than try to invest more time in this particular one.

So my plan is:

1. Finish the linear extrusion method:
    1. creates a series of points from every vertex of the polygon(s) and extends them upward. 
    2. Call the Rivara LEB triangulate method on each new face. (eventually this can be done concurrently)
    3. Triangulate the top and bottom.
2. Write a couple basic tests
3. Import the library into OpenSCAD and call it at the appropriate location
4. Make the PR
5. Others can try it

**Comment by coryrc** [comment]
API advice: 

1. I think I should support some sort of progress report. `std::ostream`? A callback std::function? Other suggestions? (Right now I just print things to stderr, but that probably shouldn't be in the final, right?)
2. I should provide a way to cancel the ongoing operation; pass in a `std:: atomic<bool>`?

**Comment by jordanbrown0** [comment]
If a single operation is so expensive that progress reporting seems desirable, I have to really wonder whether it's worth it.

That said... you could look at how the existing render progress reporting works. I don't know whether preview can use that mechanism, though.

An output stream of whatever sort does not seem appropriate.  The most useful place for progress reporting is in the GUI, where an output stream is somewhere between useless (for stdout/stderr) and awkward to consume (for a stream to be consumed by a different thread).

**Comment by coryrc** [comment]
The problem is inherently very complicated. Without feedback people can think it's stuck. The only other comparable is Minkowski. Every other operation is fairly well bounded and so the progress bar does advance regularly.

I remember now, the main thread must call the output to print to the console: 

https://github.com/openscad/openscad/blob/fa8ff8916a9090d9bc64e9d3ad2725ba1aa74dce/src/gui/MainWindow.cc#L2106

In Go it's a very typical pattern to do something like: 

```
go runOperationInAGreenThread(channel);
while(! done) {
  switch(channel) {
    case message: consoleOutput(message);
    case bool: done=true;
  }
}
```

Unfortunately the caller is not in the GUI code, so something would need to multiplex it.

ostream also works because v1 can just pass in std:cerr and if someone wants console messages they can add them later.

**Comment by jordanbrown0** [comment]
The question is whether the result is worth the processing time.  (IIRC, this performance hit is not for the circular cases, but rather is for linear extrusion with twist?)  Would anybody ever notice the difference between a truly correct (and computationally expensive) calculation, and one that is not quite as "correct" but is near-zero cost?

With respect to progress reporting... I think it would be actively bad to further clutter stdout/stderr, and for the GUI stdout/stderr are almost completely useless.

**Comment by coryrc** [comment]
> (IIRC, this performance hit is not for the circular cases, but rather is for linear extrusion with twist?)

Only extrusion with twist *and* scaling. Simple solid geometry are constant-time calculations (sphere, cylinder).

> Would anybody ever notice the difference between a truly correct (and computationally expensive) calculation, and one that is not quite as "correct" but is near-zero cost?

The latter is impossible for the edge cases. You're effectively just asking for the default settings, which is just to divide the rows and columns into a grid, and has no guarantees at all about error. Anything better is computationally demanding to varying degree.
